crypto/bio/bf_readbuff.c

```diff
@@ -0,0 +1,265 @@
+        if (i <= 0) {
+            BIO_copy_next_retry(b);
+            if (i < 0)
+                return ((num > 0) ? num : i);
+            else
+                return num; /* i == 0 */
+        }
+        ctx->ibuf_len = i;
+    }
+}
+
+static int readbuffer_write(BIO *b, const char *in, int inl)
+{
+    return 0;
+}
+static int readbuffer_puts(BIO *b, const char *str)
+{
+    return 0;
+}
+
+static long readbuffer_ctrl(BIO *b, int cmd, long num, void *ptr)
+{
+    BIO_F_BUFFER_CTX *ctx;
+    long ret = 1, sz;
+
+    ctx = (BIO_F_BUFFER_CTX *)b->ptr;
+
+    switch (cmd) {
+    case BIO_CTRL_EOF:
+        if (ctx->ibuf_len > 0)
+            return 0;
+        if (b->next_bio == NULL)
+            return 1;
+        ret = BIO_ctrl(b->next_bio, cmd, num, ptr);
+        break;
+
+    case BIO_C_FILE_SEEK:
+    case BIO_CTRL_RESET:
+        sz = ctx->ibuf_off + ctx->ibuf_len;
+        /* Assume it can only seek backwards */
+        if (num < 0 || num > sz)
+            return 0;
+        ctx->ibuf_off = num;
+        ctx->ibuf_len = sz - num;
+        break;
+
+    case BIO_C_FILE_TELL:
+    case BIO_CTRL_INFO:
+        ret = (long)ctx->ibuf_off;
+        break;
+    case BIO_CTRL_PENDING:
+        ret = (long)ctx->ibuf_len;
+        if (ret == 0) {
+            if (b->next_bio == NULL)
+                return 0;
+            ret = BIO_ctrl(b->next_bio, cmd, num, ptr);
+        }
+        break;
+    case BIO_CTRL_DUP:
+    case BIO_CTRL_FLUSH:
+        ret = 1;
+        break;
+    default:
+        ret = 0;
+        break;
+    }
+    return ret;
+}
+
+static long readbuffer_callback_ctrl(BIO *b, int cmd, BIO_info_cb *fp)
+{
+    if (b->next_bio == NULL)
+        return 0;
+    return BIO_callback_ctrl(b->next_bio, cmd, fp);
+}
+
+static int readbuffer_gets(BIO *b, char *buf, int size)
+{
+    BIO_F_BUFFER_CTX *ctx;
+    int num = 0, i, flag;
+    char *p;
+
+    ctx = (BIO_F_BUFFER_CTX *)b->ptr;
+    size--;                     /* reserve space for a '\0' */
+    BIO_clear_retry_flags(b);
+
+    for (;;) {
+        if (ctx->ibuf_len > 0) {
+            p = &(ctx->ibuf[ctx->ibuf_off]);
+            flag = 0;
+            for (i = 0; (i < ctx->ibuf_len) && (i < size); i++) {
+                *(buf++) = p[i];
+                if (p[i] == '\n') {
+                    flag = 1;
+                    i++;
+                    break;
+                }
+            }
+            num += i;
+            size -= i;
+            ctx->ibuf_len -= i;
+            ctx->ibuf_off += i;
+            if (flag || size == 0) {
+                *buf = '\0';
+                return num;
+            }
+        } else {
+            /* read another chunk - resize if we have to */
+            if (!readbuffer_resize(ctx, DEFAULT_BUFFER_SIZE))
+                return 0;
+
+            /* Do some buffering by reading from the next bio */
+            i = BIO_read(b->next_bio, ctx->ibuf + ctx->ibuf_off,
+                         ctx->ibuf_size - ctx->ibuf_off);
```

Comment:
This is wrong. You do not want to do any readahead in the BIO_f_readbuffer(). This should read from the next bio just byte by byte. Otherwise you will too eagerly read too much from the stdin and break cases like reading first a key and then a certificate from stdin that was supported before by the pkcs12 app.

Reply:
Are you just talking about the gets() case??
This code path wont be used by the decoder now - so this is not really an issue here.
BIO_tell() will work for input redirection so it should not even get here for that case now.

In the initial pass it did get here since I was unconditionally adding the bio if stdin was being used.


will work for input redirection

Reply:
Yes, I am talking here about the gets. It definitely has an effect as the pem2der decoder uses BIO_gets() and so you'll pull the whole DEFAULT_BUFFER_SIZE from the stdin on first BIO_gets() call on the stdin BIO. I straced pkcs12 command and it will read 4096 bytes from stdin with your patch at once and when I change the `ctx->ibuf_size - ctx->ibuf_off` to just 1, it will read from stdin byte by byte. There is still something more though to fully fix the pkcs12 stdin input.

Reply:
Now with both commits from #14449 the PKCS12 use case is fixed as well. So please change this to read just one byte here.

Reply:
Can you give an example that worked with 1.1.1?

Reply:
I am not convinced that byte by byte reading is the best way to do this.

Reply:
The example working in 1.1.1 is #13826. If you actually strace this command in 1.1.1 you'll see that the stdin is read byte by byte as well.

Reply:
I decided to delegate gets() to the next_bio in the chain by calling its gets(). I think this is better behaviour.
The file or fd bio will then just behave like they are being called directly.. Maybe this fixes the issue you had with testing EOF also..

Reply:
That should work and according to my testing it really works! Nice trick.